File: frontend/fix_maps.js
const fs = require('fs');
const path = require('path');

const directory = 'src';

const pattern1 = /\.then\(r => r\.data\?\.data\?\.results \|\| r\.data\?\.results \|\| r\.data\?\.data \|\| r\.data \|\| \[\]\)/g;
const replacement1 = `.then(r => { const res = r.data?.data?.results || r.data?.results || r.data?.data || r.data; return Array.isArray(res) ? res : [] })`;

const pattern2 = /\.then\(r => r\.data\?\.results \|\| r\.data\?\.data \|\| r\.data \|\| \[\]\)/g;
const replacement2 = `.then(r => { const res = r.data?.results || r.data?.data || r.data; return Array.isArray(res) ? res : [] })`;

function fixIndex(content) {
    let c = content;
    c = c.replace(/const d = r\.data\?\.data \?\? r\.data\s+setHeroProducts\(d \|\| \[\]\)/g, "const d = r.data?.data ?? r.data\n        setHeroProducts(Array.isArray(d) ? d : [])");
    c = c.replace(/const d = r\.data\?\.data \?\? r\.data\s+setCategories\(d \|\| \[\]\)/g, "const d = r.data?.data ?? r.data\n        setCategories(Array.isArray(d) ? d : [])");
    c = c.replace(/const d = r\.data\?\.data \?\? r\.data\s+setMarketplace\(d \|\| \[\]\)/g, "const d = r.data?.data ?? r.data\n        setMarketplace(Array.isArray(d) ? d : [])");
    c = c.replace(/const d = r\.data\?\.data \?\? r\.data\s+setStores\(d \|\| \[\]\)/g, "const d = r.data?.data ?? r.data\n        setStores(Array.isArray(d) ? d : [])");
    return c;
}

function walkSync(dir, filelist = []) {
    const files = fs.readdirSync(dir);
    for (const file of files) {
        const filepath = path.join(dir, file);
        if (fs.statSync(filepath).isDirectory()) {
            filelist = walkSync(filepath, filelist);
        } else {
            if (filepath.endsWith('.tsx') || filepath.endsWith('.ts')) {
                filelist.push(filepath);
            }
        }
    }
    return filelist;
}

const files = walkSync(directory);
for (const file of files) {
    let content = fs.readFileSync(file, 'utf8');
    const orig = content;
    
    content = content.replace(pattern1, replacement1);
    content = content.replace(pattern2, replacement2);
    content = fixIndex(content);
    
    if (orig !== content) {
        fs.writeFileSync(file, content, 'utf8');
        console.log(`Fixed ${file}`);
    }
}
console.log('Done');

File: frontend/fix_maps.py
import os
import re

directory = r"c:\Parsonal Website\CampusHat\frontend\src"

pattern1 = re.compile(r"\.then\(r => r\.data\?\.data\?\.results \|\| r\.data\?\.results \|\| r\.data\?\.data \|\| r\.data \|\| \[\]\)")
replacement1 = r".then(r => { const res = r.data?.data?.results || r.data?.results || r.data?.data || r.data; return Array.isArray(res) ? res : [] })"

pattern2 = re.compile(r"\.then\(r => r\.data\?\.results \|\| r\.data\?\.data \|\| r\.data \|\| \[\]\)")
replacement2 = r".then(r => { const res = r.data?.results || r.data?.data || r.data; return Array.isArray(res) ? res : [] })"

pattern3 = re.compile(r"const d = r\.data\?\.data \?\? r\.data\n\s+set.+?\(d \|\| \[\]\)")

# Specifically for page.tsx index
def fix_index(content):
    c = content.replace("const d = r.data?.data ?? r.data\n        setHeroProducts(d || [])", "const d = r.data?.data ?? r.data\n        setHeroProducts(Array.isArray(d) ? d : [])")
    c = c.replace("const d = r.data?.data ?? r.data\n        setCategories(d || [])", "const d = r.data?.data ?? r.data\n        setCategories(Array.isArray(d) ? d : [])")
    c = c.replace("const d = r.data?.data ?? r.data\n        setMarketplace(d || [])", "const d = r.data?.data ?? r.data\n        setMarketplace(Array.isArray(d) ? d : [])")
    c = c.replace("const d = r.data?.data ?? r.data\n        setStores(d || [])", "const d = r.data?.data ?? r.data\n        setStores(Array.isArray(d) ? d : [])")
    return c

for root, _, files in os.walk(directory):
    for file in files:
        if file.endswith(".tsx") or file.endswith(".ts"):
            path = os.path.join(root, file)
            with open(path, "r", encoding="utf-8") as f:
                content = f.read()
            
            orig = content
            content = pattern1.sub(replacement1, content)
            content = pattern2.sub(replacement2, content)
            content = fix_index(content)
            
            if orig != content:
                with open(path, "w", encoding="utf-8") as f:
                    f.write(content)
                print(f"Fixed {path}")

print("Done")
